File: client/src/types/filter/repository.ts
import { Filter } from './filter';
import { Pagination } from './pagination';

export interface ListRequest {
  filters: Filter[];
  pagination: Pagination;
}

export interface ListResponse<T> {
  data: T[];
  pagination: Pagination;
}

export abstract class ModelRepository<T> {
  abstract list(req: ListRequest, eagerLoads?: string[]): Promise<ListResponse<T>>;
  abstract findAll(eagerLoads?: string[]): Promise<T[]>;
  abstract findById(id: string, eagerLoads?: string[]): Promise<T | null>;
  abstract create(entity: T): Promise<void>;
  abstract update(entity: T): Promise<void>;
  abstract delete(entity: T): Promise<void>;
}

File: core/internal/repository/repository.go
// repository/model_repository.go
package repository

import (
	"gorm.io/gorm"
)

type ModelRepository[T any] struct {
	db *gorm.DB
}

func NewModelRepository[T any](db *gorm.DB) *ModelRepository[T] {
	return &ModelRepository[T]{db: db}
}

func (r *ModelRepository[T]) preload(db *gorm.DB, eagerLoads []string) *gorm.DB {
	for _, relation := range eagerLoads {
		db = db.Preload(relation)
	}
	return db
}
func (r *ModelRepository[T]) FindAll(eagerLoads []string) ([]T, error) {
	var entities []T
	query := r.preload(r.db, eagerLoads)
	if err := query.Find(&entities).Error; err != nil {
		return nil, err
	}
	return entities, nil
}

func (r *ModelRepository[T]) FindByID(id string, eagerLoads []string) (T, error) {
	var entity T
	query := r.preload(r.db, eagerLoads)
	if err := query.First(&entity, "id = ?", id).Error; err != nil {
		return entity, err
	}
	return entity, nil
}

func (r *ModelRepository[T]) Create(entity *T) error {
	return r.db.Create(entity).Error
}

func (r *ModelRepository[T]) Update(entity *T) error {
	return r.db.Save(entity).Error
}

func (r *ModelRepository[T]) Delete(entity *T) error {
	return r.db.Delete(entity).Error
}
